modules/transport-netty4/src/internalClusterTest/java/org/elasticsearch/http/netty4/Netty4ChunkedContinuationsIT.java

```diff
@@ -643,10 +643,10 @@ public void close() {
 
                             @Override
                             public void accept(RestChannel channel) {
-                                localRefs.mustIncRef();
                                 client.execute(TYPE, new Request(), new RestActionListener<>(channel) {
                                     @Override
                                     protected void processResponse(Response response) {
+                                        localRefs.mustIncRef();
```

Comment:
After looking at a failed build scan with the logging turned on, it became apparent that the unbalanced sequence of events went like (most recent first):

1. BaseRestHandler closing the RestChannelConsumer (decRef)
	- at `Netty4ChunkedContinuationsIT:641`
2. Closing the resource tracker (decRef)
	- at `Netty4ChunkedContinuationsIT:324`
3. RestChannelConsumer.accept (incRef)
	- at `Netty4ChunkedContinuationsIT:646`
4. prepareRequest (incRef)
	- at `Netty4ChunkedContinuationsIT:637`
5. created (incRef implicit)
	- at `Netty4ChunkedContinuationsIT.java:322`

Which suggests to me the following sequence of events:
1. `RestChannelConsumer.accept` is called (and `mustIncRef` is called)
2. `TransportInfiniteContinuationsAction.doExecute` is called and schedules the return of the chunked response
3. The request is cancelled, the channel is closed
4. The chunked response is returned, `RestActionListener#onResponse` fails in the call to `#ensureOpen()` and ends up calling `RestActionListener#onFailure` instead of `RestActionListener#processResponse` so the `decRef` on line 655 is never called, and we have our inbalance.

### Options for fixing
If we move the `mustIncRef` such that it only occurs when we receive the `InfiniteContinuationsPlugin.Response` from the action (as above), the test shouldn't be susceptible to such a failure, but it does indicate that the `Response` object can be created but never closed. I'm not sure if this is an actual bug. `Response` type parameter doesn't have any type constraints in `RestActionListener` so it's not necessarily `Releasable`. I'm not clear on what the contract is here.

Reply:
I think the new positioning of `mustIncRef` is OK, because instead of

"if we accepted the request" -> "then we close the chunked response"

we now assert

"if we created the chunked response" -> "then we close the chunked response"

But I may have missed something.

Reply:
The other plugins (`YieldsContinuationsPlugin` at least) seem to have the same issue, but I guess it doesn't trigger because we wait for the request to complete when calling those.

Reply:
Good job on tracking this down. I can confirm the test can fail with following diff
```diff
--- a/modules/transport-netty4/src/internalClusterTest/java/org/elasticsearch/http/netty4/Netty4ChunkedContinuationsIT.java
+++ b/modules/transport-netty4/src/internalClusterTest/java/org/elasticsearch/http/netty4/Netty4ChunkedContinuationsIT.java
@@ -645,6 +645,12 @@ public class Netty4ChunkedContinuationsIT extends ESNetty4IntegTestCase {
                             public void accept(RestChannel channel) {
                                 localRefs.mustIncRef();
                                 client.execute(TYPE, new Request(), new RestActionListener<>(channel) {
+                                    @Override
+                                    protected void ensureOpen() {
+                                        safeSleep(500);
+                                        super.ensureOpen();
+                                    }
+
                                     @Override
                                     protected void processResponse(Response response) {
                                         channel.sendResponse(RestResponse.chunked(RestStatus.OK, response.getResponseBodyPart(), () -> {
```

I think this is the right fix we are doing the same thing in other `RestActionListener` [implementation](https://github.com/elastic/elasticsearch/blob/f0afd9165064edf9d3e18a5b960e57bf96de71f2/server/src/main/java/org/elasticsearch/rest/action/RestChunkedToXContentListener.java#L44). 👍

Reply:
Good catch, the analysis looks right to me. However I would still prefer that we didn't just leak the response listener in this case, that's not how production code should behave, and we should be able to assert that it's always completed. I opened https://github.com/elastic/elasticsearch/pull/110309 with an alternative fix which I would prefer.